File: drone_interface/drone_interface/px4_odom_bridge.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, qos_profile_sensor_data

from px4_msgs.msg import VehicleOdometry
from nav_msgs.msg import Odometry
from geometry_msgs.msg import TransformStamped
from builtin_interfaces.msg import Time
from tf2_ros import TransformBroadcaster

import numpy as np
from scipy.spatial.transform import Rotation as R


class Px4OdomBridge(Node):
    def __init__(self):
        super().__init__('px4_odom_bridge')

        odom_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )

        self.sub = self.create_subscription(
            VehicleOdometry,
            '/fmu/out/vehicle_odometry',
            self.callback,
            qos_profile_sensor_data
        )

        self.odom_pub = self.create_publisher(
            Odometry, '/odom', odom_qos
        )

        self.tf_broadcaster = TransformBroadcaster(self)

        self.use_sim_time = self.get_parameter('use_sim_time').value
        self.get_logger().info(
            f"Stamping with: {'sim /clock' if self.use_sim_time else 'PX4 time'}"
        )

    # ────────────────────────────────────────────────
    # Frame conversions
    # ────────────────────────────────────────────────

    @staticmethod
    def ned_to_enu_position(p):
        """NED → ENU position"""
        return [p[1], p[0], -p[2]]

    # @staticmethod
    # def aircraft_to_baselink_orientation(quat):
    #     """Rotate quaternion from PX4 aircraft frame (FRD) to ROS base_link frame (FLU)"""
    #     r = R.from_quat([quat[1], quat[2], quat[3], quat[0]])  # x,y,z,w
    #     r_rot = R.from_euler('x', np.pi)  # 180° around X
    #     r_out = r_rot * r
    #     q_out = r_out.as_quat()
    #     return np.array([q_out[3], q_out[0], q_out[1], q_out[2]])  # w,x,y,z

    # @staticmethod
    # def ned_to_enu_orientation(quat):
    #     """Rotate quaternion from NED → ENU frame"""
    #     r = R.from_quat([quat[1], quat[2], quat[3], quat[0]])
    #     # NED→ENU: +90° Z, +180° X
    #     r_rot = R.from_euler('z', np.pi/2) * R.from_euler('x', np.pi)
    #     r_out = r_rot * r
    #     q_out = r_out.as_quat()
    #     return np.array([q_out[3], q_out[0], q_out[1], q_out[2]])

    # @staticmethod
    # def px4_to_ros_orientation(quat_px4):
    #     """Full conversion: Aircraft frame → Base_link → NED→ENU"""
    #     q_baselink = Px4OdomBridge.aircraft_to_baselink_orientation(quat_px4)
    #     q_ros = Px4OdomBridge.ned_to_enu_orientation(q_baselink)
    #     return q_ros

    @staticmethod
    def px4_to_ros_orientation(quat_px4):
        r = R.from_quat([quat_px4[1], quat_px4[2], quat_px4[3], quat_px4[0]])

        # Only Z rotation needed after simplification
        r_rot = R.from_euler('z', np.pi/2)

        r_out = r_rot * r
        q_out = r_out.as_quat()

        return np.array([q_out[3], q_out[0], q_out[1], q_out[2]])
    
    @staticmethod
    def frd_to_flu_angular_velocity(w):
        """Convert angular velocity FRD → FLU"""
        rot = R.from_euler('x', np.pi)
        return rot.apply(w)

    # ────────────────────────────────────────────────
    # Callback
    # ────────────────────────────────────────────────

    def callback(self, msg: VehicleOdometry):

        # ── Time ─────────────────────────────────────
        if self.use_sim_time:
            stamp = self.get_clock().now().to_msg()
        else:
            stamp = Time(
                sec=int(msg.timestamp_sample // 1_000_000),
                nanosec=int((msg.timestamp_sample % 1_000_000) * 1000)
            )

        # ── Position ─────────────────────────────────
        px, py, pz = self.ned_to_enu_position(msg.position)

        # ── Orientation ──────────────────────────────
        qw, qx, qy, qz = self.px4_to_ros_orientation(msg.q)

        # ── Linear velocity ──────────────────────────
        vx, vy, vz = self.ned_to_enu_position(msg.velocity)

        # ── Angular velocity ─────────────────────────
        ang = np.array(msg.angular_velocity)
        wx, wy, wz = self.frd_to_flu_angular_velocity(ang)

        # ── Odometry message ─────────────────────────
        odom = Odometry()
        odom.header.stamp = stamp
        odom.header.frame_id = 'odom'
        odom.child_frame_id = 'base_link'

        odom.pose.pose.position.x = float(px)
        odom.pose.pose.position.y = float(py)
        odom.pose.pose.position.z = float(pz)

        odom.pose.pose.orientation.w = float(qw)
        odom.pose.pose.orientation.x = float(qx)
        odom.pose.pose.orientation.y = float(qy)
        odom.pose.pose.orientation.z = float(qz)

        odom.twist.twist.linear.x = float(vx)
        odom.twist.twist.linear.y = float(vy)
        odom.twist.twist.linear.z = float(vz)

        odom.twist.twist.angular.x = float(wx)
        odom.twist.twist.angular.y = float(wy)
        odom.twist.twist.angular.z = float(wz)

        # Covariance
        odom.pose.covariance[0]  = 0.01
        odom.pose.covariance[7]  = 0.01
        odom.pose.covariance[14] = 0.01
        odom.pose.covariance[21] = 0.1
        odom.pose.covariance[28] = 0.1
        odom.pose.covariance[35] = 0.1

        odom.twist.covariance[0]  = 0.1
        odom.twist.covariance[7]  = 0.1
        odom.twist.covariance[14] = 0.1
        odom.twist.covariance[21] = 0.2
        odom.twist.covariance[28] = 0.2
        odom.twist.covariance[35] = 0.2

        self.odom_pub.publish(odom)

        # ── TF ──────────────────────────────────────
        t = TransformStamped()
        t.header.stamp = stamp
        t.header.frame_id = 'odom'
        t.child_frame_id = 'base_link'

        t.transform.translation.x = float(px)
        t.transform.translation.y = float(py)
        t.transform.translation.z = float(pz)
        t.transform.rotation = odom.pose.pose.orientation

        self.tf_broadcaster.sendTransform(t)


def main(args=None):
    rclpy.init(args=args)
    node = Px4OdomBridge()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: drone_interface/src/px4_odom_bridge.cpp
#include <rclcpp/rclcpp.hpp>
#include <px4_msgs/msg/vehicle_odometry.hpp>
#include <nav_msgs/msg/odometry.hpp>

#include <px4_ros_com/frame_transforms.h>

#include <tf2_ros/transform_broadcaster.h>
#include <geometry_msgs/msg/transform_stamped.hpp>

using namespace std::placeholders;
using px4_msgs::msg::VehicleOdometry;

class OdomBridge : public rclcpp::Node
{
public:
    OdomBridge() : Node("px4_odom_bridge")
    {
        rmw_qos_profile_t qos_profile = rmw_qos_profile_sensor_data;
        auto qos = rclcpp::QoS(rclcpp::QoSInitialization(qos_profile.history, 5), qos_profile);

        sub_ = this->create_subscription<VehicleOdometry>(
            "/fmu/out/vehicle_odometry",
            qos,
            std::bind(&OdomBridge::callback, this, _1));

        pub_ = this->create_publisher<nav_msgs::msg::Odometry>("/odom", 10);

        tf_broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
    }

private:
    void callback(const VehicleOdometry::SharedPtr msg)
    {
        using namespace px4_ros_com::frame_transforms;

        // Position
        Eigen::Vector3d pos_ned(msg->position[0], msg->position[1], msg->position[2]);
        Eigen::Vector3d pos_enu = ned_to_enu_local_frame(pos_ned);

        // Velocity
        Eigen::Vector3d vel_ned(msg->velocity[0], msg->velocity[1], msg->velocity[2]);
        Eigen::Vector3d vel_enu = ned_to_enu_local_frame(vel_ned);

        // Orientation
        Eigen::Quaterniond q_ned(msg->q[0], msg->q[1], msg->q[2], msg->q[3]);
        Eigen::Quaterniond q_enu = px4_to_ros_orientation(q_ned);

        nav_msgs::msg::Odometry odom;

        // Fetch synced simulation/clock time context
        rclcpp::Time current_time = this->get_clock()->now();
        // rclcpp::Time current_time = rclcpp::Time(msg->timestamp * 1000); // PX4 is in microseconds
        
        odom.header.stamp = current_time;
        odom.header.frame_id = "odom";
        odom.child_frame_id = "base_link";

        // Pose
        odom.pose.pose.position.x = pos_enu.x();
        odom.pose.pose.position.y = pos_enu.y();
        odom.pose.pose.position.z = pos_enu.z();

        odom.pose.pose.orientation.w = q_enu.w();
        odom.pose.pose.orientation.x = q_enu.x();
        odom.pose.pose.orientation.y = q_enu.y();
        odom.pose.pose.orientation.z = q_enu.z();

        // Twist
        odom.twist.twist.linear.x = vel_enu.x();
        odom.twist.twist.linear.y = vel_enu.y();
        odom.twist.twist.linear.z = vel_enu.z();

        pub_->publish(odom);

        // -------------------------------------------------------------
        // TRANSFORM 1: odom -> base_link
        // -------------------------------------------------------------
        geometry_msgs::msg::TransformStamped t_odom;

        t_odom.header.stamp = current_time;
        t_odom.header.frame_id = "odom";
        t_odom.child_frame_id = "base_link";

        t_odom.transform.translation.x = pos_enu.x();
        t_odom.transform.translation.y = pos_enu.y();
        t_odom.transform.translation.z = pos_enu.z();

        t_odom.transform.rotation.w = q_enu.w();
        t_odom.transform.rotation.x = q_enu.x();
        t_odom.transform.rotation.y = q_enu.y();
        t_odom.transform.rotation.z = q_enu.z();

        tf_broadcaster_->sendTransform(t_odom);

        // -------------------------------------------------------------
        // TRANSFORM 2: base_link -> laser (Incorporated Static TF)
        // -------------------------------------------------------------
        geometry_msgs::msg::TransformStamped t_laser;

        t_laser.header.stamp = current_time; // Strictly synced to the same clock
        t_laser.header.frame_id = "base_link";
        t_laser.child_frame_id = "laser";

        // Adjust physical LiDAR offsets here if needed (e.g., raised 10cm up)
        t_laser.transform.translation.x = 0.0;
        t_laser.transform.translation.y = 0.0;
        t_laser.transform.translation.z = 0.1;

        t_laser.transform.rotation.w = 1.0;
        t_laser.transform.rotation.x = 0.0;
        t_laser.transform.rotation.y = 0.0;
        t_laser.transform.rotation.z = 0.0;

        tf_broadcaster_->sendTransform(t_laser);
    }

    rclcpp::Subscription<VehicleOdometry>::SharedPtr sub_;
    rclcpp::Publisher<nav_msgs::msg::Odometry>::SharedPtr pub_;
    std::unique_ptr<tf2_ros::TransformBroadcaster> tf_broadcaster_;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<OdomBridge>());
    rclcpp::shutdown();
    return 0;
}